modules/dists/SparseBlockDist.chpl

```diff
@@ -117,6 +119,102 @@ class SparseBlockDom: BaseSparseDom {
     dsiAdd((ind,));
   }
 
+  proc dsiBulkAdd(inds: [] index(rank, idxType), isSorted=false, isUnique=false,
+      preserveInds=true){
+  
+    if !isSorted && preserveInds {
+      var _inds = inds;
+      bulkAdd_help(_inds, isSorted, isUnique); 
+    }
+    else {
+      bulkAdd_help(inds, isSorted, isUnique);
+    }
+  }
+
+  // Tried to put this record in the function and the if statement, but got a
+  // segfault from the compiler.
+  record TargetLocaleComparator {
+
+    // I wish I could return tuple from key function. There are checks to see
+    // if the type of the function is numeric/string. Maybe there should be
+    // isComparable that can do recursive checks for at least tuples of
+    // numericals/strings/comparables.
+    // In general being able to return tuples from key() can enable sorting with
+    // multiple keys.
+
+    //proc key(a: index(rank, idxType)) { 
+    //  return dist.targetLocales[dist.targetLocsIdx(a)].id; 
+    //}
+
+    proc compare(a: index(rank, idxType), b: index(rank, idxType)) {
+      var locDif = dist.targetLocales[dist.targetLocsIdx(a)].id -
+        dist .targetLocales[dist.targetLocsIdx(b)].id;
+
+      if locDif != 0 then return locDif;
+
+      if a < b then return -1;
+      if a > b then return 1;
+      return 0;
+    }
+  }
+
+  proc bulkAdd_help(inds: [] index(rank,idxType), isSorted=false, isUnique=false) {
+
+    if !isSorted {
+
+      // without _new_ krecord functions throw null deref. It doesn't seem to be 
+      // able to deref _outer_ ? I think it has something to do with memory
+      // allocation for classes vs records, but I cannot explain
+      var comp = new TargetLocaleComparator;
+      sort(inds, comparator=comp);
+    }
```

Comment:
The case for isSorted doesn't quite make sense to me. if !isSorted, we sort by a special comparator that first puts the destination locale. Otherwise, it would be sorted, what, numerically by index? How would that affect the code in bulkAddMemoize?